src/fzx/tui/color.hpp

```diff
@@ -0,0 +1,89 @@
+#pragma once
+#include "fzx/macros.hpp"
+#include <cstdint>
+#include <variant>
+#include <string_view>
+#include <fmt/core.h>
+
+namespace fzx {
+constexpr static uint8_t parseNibble(char c)
+{
+  ASSERT((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f'));
+  if (c >= '0' && c <= '9') {
+    return c - '0';
+  } else {
+    return c - 'a' + 0xa;
+  }
+}
+
+constexpr static uint8_t parseByte(std::string_view byte)
+{
+  ASSERT(byte.size() == 2);
+  return parseNibble(byte[0]) * 16 + parseNibble(byte[1]);
+}
+
+enum TermColorCode : uint8_t {
+  Black = 0,
+  Red = 1,
+  Green = 2,
+  Yellow = 3,
+  Blue = 4,
+  Purple = 5,
+  Cyan = 6,
+  White = 7,
+  Default = 9,
+};
+
+struct TermColor
+{
+  TermColorCode mCode;
+  bool mBright;
+  constexpr TermColor(TermColorCode code, bool bright) : mCode(code), mBright(bright) { }
+  constexpr TermColor(TermColorCode code) : mCode(code), mBright(false) { }
+};
+
+struct TrueColor
+{
+  uint8_t mRed, mGreen, mBlue;
+  constexpr TrueColor(std::string_view hexString)
+    : mRed(parseByte(hexString.substr(1, 2)))
+    , mGreen(parseByte(hexString.substr(3, 2)))
+    , mBlue(parseByte(hexString.substr(5, 2)))
+  {
+    ASSERT(hexString.size() == 7);
+    ASSERT(hexString.front() == '#');
+  }
```

Comment:
```suggestion
  uint8_t mRed = 0, mGreen = 0, mBlue = 0;
  constexpr TrueColor(std::string_view hexString)
  {
    ASSERT(hexString.size() == 7);
    ASSERT(hexString.front() == '#');
    mRed = parseByte(hexString.substr(1, 2));
    mGreen = parseByte(hexString.substr(3, 2));
    mBlue = parseByte(hexString.substr(5, 2));
  }
```

Reply:
if i do that, then it can't be constexpr without c++20, and only clang optimizes it away without the constexpr: https://godbolt.org/z/d3v16b4nv

Reply:
Maybe it's fine though, substr can throw std::out_of_range

Reply:
> then it can't be constexpr without c++20

yeah, that's why I added `= 0`, but since string_view::substr throws, it's fine